Expose name max length to CLI

BrianPugh opened the pull request:
This PR does a few things:

- Unrelated to anything else, I added `littlefs-python --version` to the cli.
- I increased `LFS_NAME_MAX` macro definition from the default `255` to `32767`.
    - Pros:
        - LittleFS supports any `name_max` configuration that is smaller than the version the driver was compiled with.
        - This effectively means that littlefs-python can support any reasonable use-case.
        - 32767 was chosen because it's the Windows maximum path length; I doubt any reasonable use-case would even approach this value.
    - Cons:
        - The `lfs_info` struct is now `32512` larger in `littlefs-python`. I don't think any reasonable user would care about this.
        - The `lfs_migrate` in `littlefs-python` also now uses `32512` more memory. This is a rarely used function, and also no reasonable user would care about this increase.
    - Additional Info:
        - LittleFS stores the config's `name_max` in it's superblock.
        - A LittleFS driver (understandably) cannot mount (it will return an error) a partition who has a superblock `name_max` larger than the `LFS_NAME_MAX` the driver was compiled with.
- I also exposed setting the `name_max` parameter in the CLI via `--name-max=255`.

I should have given more thought to this in the original CLI PR, but better late than never 😅 

This is the final blocker for the related esp_littlefs PR: https://github.com/joltwallet/esp_littlefs/pull/140

ianlevesque commented:
Looks like there's a bug here:

ianlevesque commented:
This seems to have a bug:

```
Traceback (most recent call last):
  File "build/littlefs_py_venv/lib/python3.10/site-packages/littlefs/__main__.py", line 63, in validate_args
    print(f"  Max Name Len: {args.max_name_len:6d}")
AttributeError: 'Namespace' object has no attribute 'max_name_len'
```

ianlevesque commented on src/littlefs/__main__.py:
```diff
@@ -50,18 +57,19 @@ def validate_args(parser: argparse.ArgumentParser, args: argparse.Namespace):
+        print(f"  Block Size:   {args.block_size:6d}  /  0x{args.block_size:X}")
+        print(f"  Image Size:   {args.fs_size:6d}  /  0x{args.fs_size:X}")
+        print(f"  Block Count:  {args.block_count:6d}")
+        print(f"  Max Name Len: {args.max_name_len:6d}")
```
`print(f"  Max Name Len: {args.name_max:6d}")`

BrianPugh replied:
whoops! missed that one; fixed in aafeb58

jrast commented:
> LittleFS stores the config's name_max in it's superblock.

What are the consequences of this? Most embedded systems will use a much lower limit for this.

BrianPugh commented:
* In this PR, the actual default `name_max` default is still 255 (in the `class LFSConfig`). The `LFS_NAME_MAX` macro definition controls (both of which, **only** impact littlefs-python memory efficiency):
    * The upper limit of what `cfg->name_max` can be.
    * The size of the `struct lfs_info`.
* Because the default is still 255, by default the stored value in the superblock will be 255. The parameter provided by `--name-max` in the cli influences this. For example, `--name-max=64` would mean that a `name_max=64` would be stored in the superblock.
* For embedded consumers:
    * Fact: `cfg->name_max` must be less than or equal to `LFS_NAME_MAX`. This is a different `LFS_NAME_MAX` from littlefs-python's described above.
    * Everything mounts fine if `cfg->name_max` is greater-than-or-equal-to the superblock's stored `name_max`.
    * Mounting will return `LFS_ERR_INVAL` if the superblock's `name_max` exceeds `cfg->name_max`.
    
    
So, to directly answer your question, setting the high `LFS_NAME_MAX=32767` has 0 impact on the embedded target. It only results in reduced memory efficiency for a few operations in littlefs-python (of which, no one should care about). The generated littlefs binary images with this tool should be identical before/after this PR with the default values.

BrianPugh commented:
hold on, let me test this a tiny bit more

ianlevesque commented:
FWIW the description above is consistent with my experience testing this last night. I configured an esp32 target with 64 as the compile-time name-max for the littlefs library linked into the embedded target, created an image with this branch of the CLI, specifying 64 as the name-max, and it all matched up and worked fine.

BrianPugh commented:
alright, I'm happy with this. Summary of changes over the past 4 commits (mostly cosmetic, one real change):
1. To aid in debugging, I added a reasonable `__repr__` to the `LFSConfig` class.
2. I fixed a type hint in `LittleFS.__init__`.
3. I increased the spacing allowed for values when verbose printing; 6 is too small (1MB partitions are common), and I figured the largest reasonable LittleFS partition is on the order of 128MB or 256MB, so 9 spaces is sufficient.
4. This is the big one. Turns out cython disabled all asserts by default, so I fixed that in 1cd4fb45331ad7eee9680f059aa3fbe462892d58. I discovered this by setting a very small `LFS_NAME_MAX` and was shocked when code was running "fine." Whenever a user provides a `--lfs-name` that is larger than `32767`, the resulting error isn't pretty, but at least it crashes. I figure this is such a rare/unrealistic use-case that it's not worth the code clutter to try and clean up the error message.

jrast commented:
Thanks for the detailed explanation! Looks good!

jrast commented:
A new release build (version 0.7.1) is running and will be available once [this](https://github.com/jrast/littlefs-python/actions/runs/6228034501) pipline finishes.
